packages/types/src/codec/TypeRegistry.ts

```diff
@@ -0,0 +1,319 @@
+// Copyright 2017-2019 @polkadot/types authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+import { assert, isString, isFunction, isUndefined } from '@polkadot/util';
+import { ArgsDef, Codec, Constructor, IMetadataEvent, RegistryTypes } from '../types';
+import Text from '../primitive/Text';
+import EnumType from './EnumType';
+import Compact from './Compact';
+import { Linkage } from './Linkage';
+import Option from './Option';
+import Struct from './Struct';
+import Tuple from './Tuple';
+import UInt from './UInt';
+import Vector from './Vector';
+import Method, { FN_UNKNOWN, IFunctionMetadata, MethodFunction, ModulesWithMethods } from '../primitive/Method';
+import { EventData } from '../type/Event';
+import { getTypeDef } from './createType';
+import flattenUniq from '../Metadata/util/flattenUniq';
+
+export enum TypeDefInfo {
+  Compact,
+  Enum,
+  Option,
+  Plain,
+  Struct,
+  Tuple,
+  Vector,
+  Linkage
+}
+
+export interface ITypeDef {
+  type: string;
+  meta: {
+    info?: TypeDefInfo;
+    keyName?: string;
+  };
+  module?: string;
+  sub?: ITypeDef | Array<ITypeDef>;
+}
+
+function isITypeDef (obj: Object): boolean {
+  const keys = Object.keys(obj);
+  return keys.includes('type') && keys.includes('meta');
+}
+
+function isPromise (promise: any): boolean {
+  return typeof promise === 'object' && typeof promise.then === 'function';
+}
+
+function toTypeDefKey (typeDef: ITypeDef) {
+  return `${typeDef.module}:${typeDef.type}`;
+}
+
+export default class TypeRegistry {
+
+  protected static _TYPE_REGISTRY?: TypeRegistry;
+  protected static TY_STACK: (TypeRegistry | undefined)[] = [];
+
+  static get TYPE_REGISTRY (): TypeRegistry {
+    assert(TypeRegistry._TYPE_REGISTRY, 'TYPE_REGISTRY is undefined, make sure it is wrapped in withRegistry()');
+    return TypeRegistry._TYPE_REGISTRY as TypeRegistry;
+  }
+
+  static withRegistry (typeRegistry: TypeRegistry, wrapFn: (...args: any[]) => any) {
+    TypeRegistry.TY_STACK.push(TypeRegistry._TYPE_REGISTRY);
+    TypeRegistry._TYPE_REGISTRY = typeRegistry;
+    const ret = wrapFn();
+    if (isPromise(ret)) {
```

Comment:
Supporting promise is probably a bad idea, the stack could be messed if two promises are resolved in a wrong order. will remove promise support later.